PEP 677: Callable Type Syntax (#2187)
python/peps: pep-0677.rst

        Pradeep Kumar Srinivasan <gohanpra@gmail.com>
Sponsor: Guido van Rossum <guido at python.org>
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: 13-Dec-2021
Python-Version: 3.11
Post-History:

Abstract
========

This PEP introduces a concise and friendly syntax for callable types,
supporting the same functionality as ``typing.Callable`` but with an
arrow syntax inspired by the syntax for typed function
signatures. This allows types like ``Callable[[int, str], bool]`` to
be written ``(int, str) -> bool``.

The proposed syntax supports all the functionality provided by
``typing.Callable`` and ``typing.Concatenate``, and is intended to
work as a drop-in replacement.

Motivation
==========

Describing Callable Signatures with ``typing.Callable``
------------------------------------------------------

One way to make code safer and easier to analyze is by making sure
that functions and classes are well-typed. In Python we have type
annotations defined by PEP 484 to provide type hints that can find
bugs as well as helping with editor tooling like tab completion,
static analysis tooling, and code review.

One of the types `defined by PEP 484

> <https://www.python.org/dev/peps/pep-0484/#callable>`_ in the is

Reviewer: ```suggestion
<https://www.python.org/dev/peps/pep-0484/#callable>`_ is
```

Reviewer: This crept back in

``typing.Callable``, which describes a callable value such as a
function or a method. It takes two parameters as inputs but with the
first parameter being either a placeholder like ``...`` or a list of

> types. For example: - ``Callable[..., int]`` indicates a funciton with

Reviewer: Should this be an indented list?

arguments. However, we decided against including these features, as
our analysis showed they are infrequently used. When they are really
needed, it is possible to type these using `Callback Protocols
<https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols>`_.

See the Rejected Alternatives section for more detailed discussion
about omitted features.

Specification
=============

Typing Behavior
---------------

Type checkers should treat the new syntax with exactly the same
semantics as ``typing.Callable``.

As such, a type checker should treat the following pairs exactly the
same::

   from typing import Awaitable, Callable, Concatenate, ParamSpec, TypeVarTuple

    P = ParamSpec("P")
    Ts = TypeVarTuple('Ts')

    f0: (int, str) -> bool
    f0: Callable[[int, str], bool]

    f1: (...) -> bool
    f1: Callable[..., bool]

    f2: async (str) -> str
    f2: Callable[[str], Awaitable[str]]

    f3: (**P) -> bool
    f3: Callable[P, bool]

    f4: (int, **P) -> bool
    f4: Callable[Concatenate[int, P], bool]

    f5: (*Ts) -> bool
    f5: Callable[[*Ts], bool]

    f6: (int, *Ts, str) -> bool
    f6: Callable[[int, *Ts, str], bool]

> 

Reviewer: Maybe add an example with no arguments: `() -> bool`. Below you say that `(,) -> bool` is an error but not what to write instead.

> Grammar and Ast

Reviewer: ```suggestion
Grammar and AST
```

Here are our proposed changes to the `Python Grammar
<https://docs.python.org/3/reference/grammar.htm>`::

    expression:
        | disjunction disjunction 'else' expression
        | callable_type_expression
        | disjunction
        | lambdef

    callable_type_expression:
        | callable_type_arguments '->' expression
        | ASYNC callable_type_arguments '->' expression

    callable_type_arguments:
        | '(' '...' [','] ')'
        | '(' callable_type_positional_argument*  ')'
        | '(' callable_type_positional_argument* callable_type_param_spec ')'

    callable_type_positional_argument:
        | !'...' expression ','
        | !'...' expression &')'

    callable_type_param_spec:
        | '**' expression ','
        | '**' expression &')'

If PEP 646 is accepted, we intend to include support for unpacked
types by modifying the grammar for
``callable_type_positional_argument`` as follows::

    callable_type_positional_argument:
        | expression ','
        | expression &')'
        | '*' expression ','
        | '*' expression &')'

Implications of the Grammar
---------------------------

Precedence of ->
~~~~~~~~~~~~~~~~

> ``->`` binds less tightly than other operators, both inside types and in function signatures::

Reviewer: Does this mean that the two examples below are equivalent? Would be good to make that explicit.

  use types with named arguments.
- Our analysis suggests that support for ``ParamSpec`` is key, but the
  scoping rules laid out in PEP 612 would have made this difficult.

Other Proposals Considered
--------------------------

Functions-as-Types
~~~~~~~~~~~~~~~~~~

An idea we looked at very early on was to `allow using functions as
types
<https://docs.google.com/document/d/1rv6CCDnmLIeDrYlXe-QcyT0xNPSYAuO1EBYjU3imU5s/edit?usp=sharing>`_.
The idea is allowing a function to stand in for its own call
signature, with roughly the same semantics as the ``__call__`` method
of Callback Protocols. Think this may be a great idea and worth its
own PEP, but that it is not a good alternative to improving the
usability of callable types:

- Using functions as types would not give us a new way of describing
  function types as first class values. Instead, they would require a
  function definition statement that effectively defines a type alias
  (much as a Callable Protocol class statement does).
- Functions-as-types would support almost exactly the same features
  that Callable Protocols do today: named, optional, and variadic args

>   as well as the ability to define overloads.

Reviewer: It's also not obvious how to support ParamSpec with this proposal.

Parenthesis-Free Syntax
~~~~~~~~~~~~~~~~~~~~~~~

We considered a parentheses-free syntax that would have been even more
concise::

    int, str -> bool

We decided against it because this is not visually as similar to
existing function header syntax. Moreover, it is visually similar to
lambdas, which bind names with no parentheses: ``lambda x, y: x ==
y``.

Introducing type-strings
~~~~~~~~~~~~~~~~~~~~~~~~

> Another idea was a new “special string” syntax an puting the type

Reviewer: ```suggestion
Another idea was adding a new “special string” syntax and putting the type
```

a very similar syntax for function type hint *comments* for use in
code that needs to work on Python 2.7. For example::

    def f(x, y):
        # type: (int, str) -> bool
        ...

At that time we used indexing operations to specify generic types like
``typing.Callable`` because we decided not to add syntax for
types. However, we have since begun to do so, e.g. with PEP 604.

**Maggie** proposed better callable type syntax as part of a larger
`presentation on typing simplifications
<https://drive.google.com/file/d/1XhqTKoO6RHtz7zXqW5Wgq9nzaEz9TXjI/view>`_
at the PyCon Typing Summit 2021.

**Steven** `brought up this proposal on typing-sig
<https://mail.python.org/archives/list/typing-sig@python.org/thread/3JNXLYH5VFPBNIVKT6FFBVVFCZO4GFR2>`. We
had several meetings to discuss alternatives, and `this presentation
<https://www.dropbox.com/s/sshgtr4p30cs0vc/Python%20Callable%20Syntax%20Proposals.pdf?dl=0>`_
led us to the current proposal.

**Pradeep** `brought this proposal to python-dev
<https://mail.python.org/archives/list/python-dev@python.org/thread/VBHJOS3LOXGVU6I4FABM6DKHH65GGCUB>`_
for feedback.

> The proposed AST and runtime behavior is explicitly designed to be

Reviewer: This paragraph feels out of place; there's already discussion of forward compatibility above.

forward-compatible with possible extensions of callable type literals
to support named, optional, and variadic parameters. We do not
currently plan to propose this extension but we consider it important
that our proposal does not rule it out. `This doc
<https://docs.google.com/document/d/1AJ0R7lgcKY0gpZbkBZRxXTvgV-OqxMYDj_JOPYMQFP8/edit>`_
has a detailed discussion of how each detail of our proposal relates
to forward compatibility.

Other Languages
---------------

Many popular programming languages use an arrow syntax similar
to the one we are proposing here

the same ``->`` arrow token we are proposing here.
almost identical to the ones we are proposing here

TypeScript
~~~~~~~~~~

In `TypeScript
<https://basarat.gitbook.io/typescript/type-system/callable#arrow-syntax>`_,
function types are expressed in a syntax almost the same as the one we
are proposing, but the arrow token is `=>` and arguments have names::

    (x: int, y: str) => bool

The names of the arguments are not actually relevant to the type. So,
for example, this is the same callable type::

    (a: int, b: str) => bool

Kotlin
~~~~~~

> Function types in `Kotlin <https://kotlinlang.org/docs/lambdas.html>`_ permits

Reviewer: ```suggestion
Function types in `Kotlin <https://kotlinlang.org/docs/lambdas.html>`_ permit
```

Our proposal closely follows syntax used by several popular languages
such as `Typescript
<https://basarat.gitbook.io/typescript/type-system/callable#arrow-syntax>`_,
`Kotlin <https://kotlinlang.org/docs/lambdas.html>`_, and `Scala
<https://docs.scala-lang.org/tour/higher-order-functions.html>`_.

Our goals are that:

- Callable types using this syntax will be easier to learn and use,
  particularly for developers with experience in other languages.
- Library authors will be more likely to use expressive types for
  callables that enable type checkers to better understand code and
  find bugs, as in the ``decorator`` example above.

Consider this simplified real-world example from a web server, written
using the existing ``typing.Callable``::

    from typing import Awaitable, Callable
    from app_logic import Response, UserSetting

    async def customize_response_for_settings(
        response: Response,
        customizer: Callable[[Response, list[UserSetting]], Awaitable[Response]]
    ) -> Response:
       ...

With our proposal, this code can be abbreviated to::

    from app_logic import Response, UserSetting

>     def make_endpoint(

Reviewer 2: Not sure why this comment is marked resolved, we still have `async def customize_response_for_settings` above and `def make_endpoint` here. AFAICS the difference is irrelevant to the example and both should be just `def make_endpoint`

nesting of square brackets - only one level, as opposed to three in
the original code.

Rationale
=========

The ``Callable`` type is widely used. For example, `as of October 2021
it was
<https://github.com/pradeep90/annotation_collector#overall-stats-in-typeshed>`_
the fifth most common complex type in typeshed, after ``Optional``,
``Tuple``, ``Union``, and ``List``.

Most of the other commonly used types have had their syntax improved
via either PEP 604 or PEP 585. ``Callable`` is used heavily enough to
similarly justify a more usable syntax.

In this proposal, we chose to support all the existing semantics of
``typing.Callable``, without adding support for new features. We took
this decision after examining how frequently each feature might be
used in existing typed and untyped open-source code. We determined
that the vast majority of use cases are covered.

We considered adding support for named, optional, and variadic
arguments. However, we decided against including these features, as
our analysis showed they are infrequently used. When they are really
needed, it is possible to type these using `Callback Protocols

> <https://mypy.readthedocs.io/en/stable/protocols.html#callback-protocols>`_.

Reviewer 3: I think neither Callback nor Protocols should be capitalized. This would have to change throughout the PEP.

See the Rejected Alternatives section for more detailed discussion

> about omitted features.

Author: Done

Specification
=============

Typing Behavior
---------------

Type checkers should treat the new syntax with exactly the same
semantics as ``typing.Callable``.

As such, a type checker should treat the following pairs exactly the
same::

    from typing import Awaitable, Callable, Concatenate, ParamSpec, TypeVarTuple

>     P = ParamSpec("P")

Reviewer 4: Feel like it is worth adding one line before each pair explaining what it is that, say, `() -> bool` means.

Five years from now, if all goes to plan, nobody will even know what `typing.Callable[[], bool]` meant 🤣  

So, it's our job to explain what it means. For complex stuff like ParamSpec, we can at least link to the PEP for more details.

    Ts = TypeVarTuple('Ts')

    f0: () -> bool
    f0: Callable[[], bool]

    f1: (int, str) -> bool
    f1: Callable[[int, str], bool]

    f2: (...) -> bool
    f2: Callable[..., bool]

    f3: async (str) -> str
    f3: Callable[[str], Awaitable[str]]

    f4: (**P) -> bool
    f4: Callable[P, bool]

    f5: (int, **P) -> bool
    f5: Callable[Concatenate[int, P], bool]

    f6: (*Ts) -> bool
    f6: Callable[[*Ts], bool]

    f7: (int, *Ts, str) -> bool
    f7: Callable[[int, *Ts, str], bool]

Grammar and AST
---------------

> The proposed new syntax can be described by these AST changes ::

Reviewer 3: Might mention this goes into Parser/Python.asdl.

    callable_type_arguments = AnyArguments
                            | ArgumentsList(expr* posonlyargs)
                            | Concatenation(expr* posonlyargs, expr param_spec)

Here are our proposed changes to the `Python Grammar
<https://docs.python.org/3/reference/grammar.htm>`::

    expression:
        | disjunction disjunction 'else' expression
        | callable_type_expression
        | disjunction
        | lambdef

    callable_type_expression:
        | callable_type_arguments '->' expression
        | ASYNC callable_type_arguments '->' expression

    callable_type_arguments:
        | '(' '...' [','] ')'
        | '(' callable_type_positional_argument*  ')'
        | '(' callable_type_positional_argument* callable_type_param_spec ')'

    callable_type_positional_argument:
        | !'...' expression ','
        | !'...' expression &')'

    callable_type_param_spec:
        | '**' expression ','
        | '**' expression &')'

If PEP 646 is accepted, we intend to include support for unpacked
types by modifying the grammar for
``callable_type_positional_argument`` as follows::

    callable_type_positional_argument:
        | expression ','

>         | expression &')'

Reviewer 3: What happened to the `!'...'` guard?

Author: Whoops - this section predates me realizing we needed that guard. Will fix!

        | '*' expression ','
        | '*' expression &')'

Implications of the Grammar
---------------------------

Precedence of ->
~~~~~~~~~~~~~~~~

``->`` binds less tightly than other operators, both inside types and
in function signatures, so the following two callable types are

> equivalent::

Reviewer 4: Can we explain it in words? This section is pretty dense and people may not know what the result is after it "binds tightly".

"..., so `(int) -> str | bool` refers to a callable that accepts `int` and returns `str | bool` rather than a union of `(int) -> str` and `bool`. That is::

    (int) -> str | bool

    # equivalent to
    (int) -> (str | bool)

    # NOT
     ((int) -> str) | bool

    (int) -> str | bool
    (int) -> (str | bool)

``->`` associates to the right, both inside types and in function

> signatures. So the following pairs are equivalent::

Reviewer 4: Same. Would appreciate mentioning it in words before introducing "associates to the right" (since that's usually in the context of arithmetic-like operators):

"``->`` associates to the right, both inside .... So, `(int) -> (str) -> bool` is a callable that accepts `int` and returns another callable `(str) -> bool`. The following pairs are equivalent::

    (int) -> (str) -> bool
    (int) -> ((str) -> bool)

    def f() -> (int, str) -> bool: pass
    def f() -> ((int, str) -> bool): pass

    def f() -> (int) -> (str) -> bool: pass
    def f() -> ((int) -> ((str) -> bool)): pass

Because operators bind more tightly than ``->``, parentheses are
required whenever an arrow type is intended to be inside an argument
to an operator like ``|``::

    (int) -> bool | () -> bool    # syntax error!

>     (int) -> bool | (() -> bool)  # okay

Reviewer 3: So that's a function from (int) to a union type, the right half of which is another function.

Maybe a more useful example would be the union of two function types?

We discussed each of these behaviors and believe they are desirable:

- Union types (represented by ``A | B`` according to PEP 604) are
  valid in function signature returns, so we need to allow operators
  in the return position for consistency.
- Given that operators bind more tightly than ``->`` it is correct
  that a type like ```bool | () -> bool`` must be a syntax error. We
  should be sure the error message is clear because this may be a
  common mistake.
- Associating ``->`` to the right, rather than requiring explicit
  parentheses, is consistent with other languages like TypeScript and
  respects the principle that valid expressions should normally be
  substitutable when possible.

``async`` Keyword
~~~~~~~~~~~~~~~~~

> All of the binding rules still work for async callable types::

Reviewer 4: ... The following pairs are equivalent::

    (int) -> async (float) -> str | bool
    (int) -> (async (float) -> (str | bool))

    def f() -> async (int, str) -> bool: pass
    def f() -> (async (int, str) -> bool): pass

    def f() -> async (int) -> async (str) -> bool: pass
    def f() -> (async (int) -> (async (str) -> bool)): pass

Trailing Commas
~~~~~~~~~~~~~~~

- Following the precedent of function signatures, putting a comma in
  an empty arguments list is illegal, ``(,) -> bool`` is a syntax
  error.
- Again following precedent, trailing commas are otherwise always

>   permitted::

Reviewer 4: "precedent" is an unfortunate choice of word here :) Maybe just "the example of function signatures"?

    ((int,) -> bool == (int) -> bool
    ((int, **P,) -> bool == (int, **P) -> bool
    ((...,) -> bool) == ((...) -> bool)

> 

Reviewer 4: Can we stick to the pair of equivalent lines formula?

Allowing trailing commas also gives autoformatters more flexibility
when splitting callable types across lines, which is always legal
following standard python whitespace rules.

Disallowing ``...`` as an Argument Type
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Under normal circumstances, any valid expression is permitted where we

> want a type annotation and ``...`` is a valid expression. This is

Reviewer 4: ^ For example, "(int, ...) -> bool".

Otherwise, the sentence is quite abstract.

never semantically valid and all type checkers would reject it, but
the grammar would allow it if we did not explicitly prevent this.

> We decided that there were compelling reasons to prevent it: - The

Reviewer 4: Formatting: needs to be a list.

PS: You can run `make rss` locally to run the CI checks (even though it wouldn't have caught this).

semantics of ``(...) -> bool`` are different from ``(T) -> bool`` for
any valid type T: ``(...)`` is a special form indicating
``AnyArguments`` whereas ``T`` is a type parameter in the arguments
list.  - ``...`` is used as a placeholder default value to indicate an
optional argument in stubs and Callback Protocols. Allowing it in the
position of a type could easily lead to confusion and possibly bugs
due to typos.

Since ``...`` is meaningless as a type and there are usability
concerns, our grammar rules it out and the following is a syntax
error::

>     (int, ...) -> bool

Reviewer 3: One more reason to disallow this is that it reminds of `tuple[int, ...]`.

Incompatibility with other possible uses of ``*`` and ``**``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The use of ``**P`` for supporting PEP 612 ``ParamSpec`` rules out any
future proposal using a bare ``**<some_type>`` to type

> ``kwargs``. This seems acceptable because:

Reviewer 3: Hm. Occasionally we hear from someone who has a TypedDict that they want to use as the type for some function's kwargs. If we wanted to support that for a function *type*'s kwargs we'd be hosed -- the obvious notation would be `(**TD) -> bool`. Or maybe that would just work, and what you're ruling out here is use of `(**int) -> bool` to type e.g. `def f(**args: int) -> bool: ...`?

Reviewer 4: Hmm... Technically, there is nothing to prevent us from supporting `**KwargsTypedDict` (or even `**int`) in the future. Type checkers could distinguish `**KwargsTypedDict` and `**P` based on the actual values of the variables. 

The main reason we have `**P` instead of a bare `P` is to make it syntactically obvious that it is a ParamSpec (like `Callable[P, R]`). Otherwise, users might have trouble realizing that P is a ParamSpec in `(Foo, P) -> int`.

We could update the PEP (in a new PR) with the above.

Author: I'm on the fence about this. Technically, there's nothing prevent us from using `**<some_type>` with divergent semantics depending on the information contained within `<some_type>`, type-checkers could handle this.

But it would be pretty unusual for python to let the semantics of the same syntax vary that much - a `ParamSpec` is a pretty different concept from typing the `kwarg` variadic argument.

My guess is that, in an extended syntax, we would work around this by requiring names for the vararg and kwarg,
so that we would use:
```
(**kwags: int) -> bool
(**kwargs: **MyTypedDict) -> bool
```

I think we should update the PEP to clarify that it's technically still possible to use `**int` and `**MyTypedDict` but I don't want to imply this is a good idea or say anything that's likely to prompt much discussion of it.

Reviewer: For what it's worth, I don't see the issue with allowing `**MyTypedDict` in the future: just like `**P`, it means "this is the type for all of the kwargs". Perhaps to keep this option open we should avoid using the name "paramspec" in the AST and runtime objects.

Reviewer 3: > But it would be pretty unusual for python to let the semantics of the same syntax vary that much

Would it, though? We have `1+2 == 3`, vs `"a" + "b" == "ab"`. We have `a[i]` (indexing list or tuple) vs. `d[key]` (lookup in dict or mapping) vs. `list[int]` (generic type). Etc. So I think there are enough precedents -- we don't need to worry about this looking odd.

- If we ever do want such a syntax, it would be clearer to require an
  argument name anyway. This would also make the type look more
  similar to a function signature. In other words, if we ever support
  typing ``kwargs`` in callable types, we would prefer ``(int,
  **kwargs: str)`` rather than ``(int, **str)``.
- PEP 646 unpacking syntax would rule out using ``*<some_type>`` for
  ``args``. The ``kwargs`` case is similar enough that this rules out
  a bare ``**<some_type>`` anyway.

Compatibility with Arrow-Based Lambda Syntax
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To the best of our knowledge there is no active discussion of
arrow-style lambda syntax that we are aware of, but it is nonetheless
worth considering what possibilities would be ruled out by adopting
this proposal.

> It would be incompatible with this proposal to adopt the same a

Reviewer 3: ```suggestion
It would be incompatible with this proposal to adopt a
```

Reviewer 4: Can you explain why? (Because `(int) -> str` is a valid expression.)

-------

That reminds me: we need to have a subsection for aliases (`Foo = (int) -> str`, generic aliases) and how a callable type behaves at runtime when used as an expression. 

@gvanrossum  might have thoughts on things we should cover (especially from python-dev's point of view).

parenthesized ``->``-based arrow syntax for lambdas, e.g.  ``(x, y) ->
x + y`` for ``lambda x, y: x + y``.

Our view is that if we want arrow syntax for lambdas in the future, it
would be a better choice to use ``=>``, e.g. ``(x, y) => x + y``.
Many languages use the same arrow token for both lambdas and callable
types, but Python is unique in that types are expressions and have to
evaluate to runtime values. Our view is that this merits using
separate tokens, and given the existing use of ``->`` for return types

> in function signatures it would be more coherent to use ``->`` for

Reviewer 3: ```suggestion
in function signatures it would be more consistent to use ``->`` for
```

- `stats on the use of the Callable type <https://github.com/pradeep90/annotation_collector#typed-projects---callable-type>`_;
- `stats on how untyped and partially-typed callbacks are actually used <https://github.com/pradeep90/annotation_collector#typed-projects---callback-usage>`_.

We decided on a simple proposal with improved syntax for the existing
``Callable`` type because the vast majority of callbacks can be correctly
described by the existing ``typing.Callable`` semantics:

- Positional parameters: By far the most important case to handle well
  is simple callable types with positional parameters, such as
  ``(int, str) -> bool``
- ParamSpec and Concatenate: The next most important feature is good
  support for PEP 612 ``ParamSpec`` and ``Concatenate`` types like
  ``(**P) -> bool`` and ``(int, **P) -> bool``. These are common
  primarily because of the heavy use of decorator patterns in python
  code.
- TypeVarTuples: The next most important feature, assuming PEP 646 is
  accepted, is for unpacked types which are common because of cases
  where a wrapper passes along ``*args`` to some other function.

Features that other, more complicated proposals would support account

> for fewer than 2% of the use cases we found. These are already

Reviewer 3: I wish we had more data to support the 2% number, since I still have a nagging doubt about it. I worry that this number will be quoted endlessly regardless of its truth value. Maybe Pradeep can publish a set of raw data sources one can use to repeat his analysis? (It would be cool if we could repeat the same analysis at some future point and see if the number is changed.)

Reviewer 4: [Guido]

I'd already shared the [script](https://github.com/pradeep90/annotation_collector#calls-to-callback-parameters) to repeat the analysis on any project from the command-line. The list of OSS projects with stats is [here](https://github.com/pradeep90/annotation_collector#stats) (raw [callables](https://github.com/pradeep90/annotation_collector/tree/master/data)). I definitely want people to repeat the analysis on their favorite code bases.

(@stroxler Could we explicitly link to the script in the resources and maybe drop a couple of lines describing the methods, the assortment of projects, and the limitations? We're referring to the stats quite a bit and people might be curious.)

[Guido]

> It would be cool if we could repeat the same analysis at some future point and see if the number is changed.

Yes, we can rerun the above script later on to see how usage has changed.

> I worry that this number will be quoted endlessly

+1. Feel like we should qualify it instead of presenting it as a bare stat, especially since people will question it. Maybe something like "X% of the cases we found in the N open-source projects we analyzed". That makes it explicit that YMMV and that this is an average.

> since I still have a nagging doubt about it

Were there particular types of projects where you feel the distribution would be different? It can help to test the hypothesis directly instead of letting the doubt nag :)

(For example, I had been curious how untyped projects used callbacks, which is why I included untyped projects like django. Same for open-source vs internal. Steven had been curious whether application code differed much from library code.)

Reviewer 3: I need to give up on fighting this on a numeric basis. I agree that it's probably little-used.

I still worry about confirmation bias, but my biggest, intuitive objection is that I feel the extended form is needed for *completeness* (i.e., there are function types whose values you can spell with a single `def` but whose type must be spelled with a callback protocol). I know it's fashionable to make fun of completionism, but I suspect that a lot of other useful features wouldn't have been in the language if we had based our decisions purely on data about current usage.

My second objection is that callback protocols are both ugly and unintuitive.

expressible using Callback Protocols, and since they are uncommon we
decided that it made more sense to move forward with a simpler syntax.

> Extended Syntax Supporting Named and Optional Arguments

Reviewer 4: nit: lower case "Supporting"

-------------------------------------------------------

Another alternative was for a compatible but more complex syntax that
could express everything in this PEP but also named, optional, and
variadic arguments. In this “extended” syntax proposal the following
types would have been equivalent::

    class Function(typing.Protocol):

>         def f(self, x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool:

Reviewer 4: `s/f/__call__/`

`s/Function/MyCallback`

            ...

>     Function = (int, y: float, *, z: bool = ..., **kwargs: str) -> bool

Reviewer 4: s/Function/MyCallback/

> Advantages of this syntax include: - Most of the advantages of the

Reviewer 4: Formatting: Needs to be a list.

proposal in this PEP (conciseness, PEP 612 support, etc) -
Furthermore, the ability to handle named, optional, and variadic
arguments

We decided against proposing it for the following reasons:

> - The implementation would have been more difficult, and usage stats

Reviewer 3: ```suggestion
- The implementation would have been more complex, and usage stats
```

Reviewer 4: ```
- Complexity: The implementation ...
```

>   demonstrate that fewer than 3% of use cases would benefit from any

Reviewer 3: Same feeling about the 3% number.

  of the added features.

> - The group that debated these proposals was split down the middle

Reviewer 4: ```
- Lack of consensus: The group that debated these proposals (typing-sig) was split ...
```

  about whether these changes are even desirable:

  - On the one hand, they make callable types more expressive. On the
    other hand, they could easily confuse users who have not read the
    full specification of callable type syntax.
  - We believe the simpler syntax proposed in this PEP, which
    introduces no new semantics and closely mimics syntax in other
    popular languages like Kotlin, Scala, and TypesScript, is much
    less likely to confuse users.

> - We intend to implement the current proposal in a way that is

Reviewer 4: ```
- Forward-compatibility:
```

>   forward-compatible with the more complicated extended syntax. If the

Reviewer 3: ```suggestion
  forward-compatible with the more complex extended syntax. If the
```

  community decides after more experience and discussion that we want

>   the additional features, they should be straightforward to propose

Reviewer 4: Ambiguous pronoun "they" (I thought it was "community").

  in the future.

> - We realized that because of overloads, it is not possible to replace

Reviewer 3: I wonder if "because of overloads" requires a complete example (e.g. an overloaded function, the observation that its type can't be expressed using current function types, and the conclusion that it must be done using a Callback Protocol). It might also be useful to explain why you can't just replace the overload with a union of function types. (Hm. Why couldn't you?)

Reviewer 4: +1 on a concrete example.

> It might also be useful to explain why you can't just replace the overload with a union of function types. (Hm. Why couldn't you?)

Overloads correspond more to an intersection of callable types. If we had a union `f: (int) -> str | ((list[int]) -> None)`, then we wouldn't be able to pass an argument to just one of the union items: `f(42)` would error because it's not a `list[int]`; `f([1, 2])` would error because it's not an `int`.

Reviewer 4: ```
- No replacement for callback protocol:
```

  all need for Callback Protocols even with an extended syntax. This
  makes us prefer proposing a simple solution that handles most use
  cases well.

We confirmed that the current proposal is forward-compatible with
extended syntax by
`implementing <https://github.com/stroxler/cpython/tree/callable-type-syntax--extended>`_
a quick-and-dirty grammar and AST on top of the grammar and AST for
the current proposal.

Syntax Closer to Function Signatures
------------------------------------

One alternative we had floated was a syntax much more similar to
function signatures.

> In this proposal, the following types would have been equivalent::

Reviewer 4: Merge into a single paragraph?

    class Function(typing.Protocol):
        def f(self, x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool:
            ...

>     Function = (x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool

Reviewer 4: Same substitutions in the previous example: `s/Function/MyCallback/g` and `s/f/__call__`

The benefits of this proposal would have included:

- Perfect syntactic consistency between signatures and callable types.
- Support for more features of function signatures (named, optional,
  variadic args) that this PEP does not support.

Key downsides that led us to reject the idea include the following:

- A large majority of use cases only use positional-only arguments,
  and this syntax would be more verbose for that use case, both
  because of requiring argument names and an explicit ``/``, for
  example ``(int, /) -> bool`` where our proposal allows ``(int) ->
  bool``
- The requirement for explicit ``/`` for positional-only arguments has
  a high risk of causing frequent bugs - which often would not be
  detected by unit tests - where library authors would accidentally
  use types with named arguments.
- Our analysis suggests that support for ``ParamSpec`` is key, but the
  scoping rules laid out in PEP 612 would have made this difficult.

Other Proposals Considered
--------------------------

Functions-as-Types
~~~~~~~~~~~~~~~~~~

> An idea we looked at very early on was to `allow using functions as
> types

Reviewer 3: I'd combine these two lines.
```suggestion
An idea we looked at very early on was to `allow using functions as types
```

<https://docs.google.com/document/d/1rv6CCDnmLIeDrYlXe-QcyT0xNPSYAuO1EBYjU3imU5s/edit?usp=sharing>`_.
The idea is allowing a function to stand in for its own call
signature, with roughly the same semantics as the ``__call__`` method
of Callback Protocols. Think this may be a great idea and worth its
own PEP, but that it is not a good alternative to improving the

> usability of callable types:

Reviewer 4: Quick example might be good (just like the other callback protocol snippets).

Can you make this sentence less informal? "Think this may be a great idea ..."

Author: Yeah, I shied away from an example because it looks kinda weird if we snake case (which is always pythonic for functions, never pythonic for types), and it also looks kinda weird if we camel case (which is always pythonic for types, never pythonic for functions). But I guess I should pick one and just look weird.

- Using functions as types would not give us a new way of describing

>   function types as first class values. Instead, they would require a

Reviewer 4: ... first-class values. That is, we would not be able to define them anywhere within a type. Instead, they would require a separate, out-of-line ...

  function definition statement that effectively defines a type alias
  (much as a Callable Protocol class statement does).

> - Functions-as-types would support almost exactly the same features

Reviewer 4: This right now reads positively. I was left a bit confused since this looked like a list of "Cons".

Maybe rephrase as "They would not support any more features than callback protocols while only being somewhat shorter."

  that Callable Protocols do today: named, optional, and variadic args
  as well as the ability to define overloads.

> Another reason we don't view functions-as-types as a good alternative

Reviewer 4: Bullet point?

a very similar syntax for function type hint *comments* for use in
code that needs to work on Python 2.7. For example::

    def f(x, y):
        # type: (int, str) -> bool
        ...

At that time we used indexing operations to specify generic types like
``typing.Callable`` because we decided not to add syntax for
types. However, we have since begun to do so, e.g. with PEP 604.

**Maggie** proposed better callable type syntax as part of a larger
`presentation on typing simplifications
<https://drive.google.com/file/d/1XhqTKoO6RHtz7zXqW5Wgq9nzaEz9TXjI/view>`_
at the PyCon Typing Summit 2021.

**Steven** `brought up this proposal on typing-sig
<https://mail.python.org/archives/list/typing-sig@python.org/thread/3JNXLYH5VFPBNIVKT6FFBVVFCZO4GFR2>`. We
had several meetings to discuss alternatives, and `this presentation
<https://www.dropbox.com/s/sshgtr4p30cs0vc/Python%20Callable%20Syntax%20Proposals.pdf?dl=0>`_
led us to the current proposal.

**Pradeep** `brought this proposal to python-dev
<https://mail.python.org/archives/list/python-dev@python.org/thread/VBHJOS3LOXGVU6I4FABM6DKHH65GGCUB>`_
for feedback.

> 

Reviewer 4: A Usage Statistics section would be good to describe the methods, projects, and [limitations](https://github.com/pradeep90/annotation_collector#limitations). We could link to it from other parts of the PEP.

TypeScript
~~~~~~~~~~

In `TypeScript
<https://basarat.gitbook.io/typescript/type-system/callable#arrow-syntax>`_,
function types are expressed in a syntax almost the same as the one we
are proposing, but the arrow token is ``=>`` and arguments have names::

    (x: int, y: str) => bool

The names of the arguments are not actually relevant to the type. So,
for example, this is the same callable type::

    (a: int, b: str) => bool

Kotlin
~~~~~~

Function types in `Kotlin <https://kotlinlang.org/docs/lambdas.html>`_ permit
an identical syntax to the one we are proposing, for example::

    (Int, String) -> Bool

It also optionally allows adding names to the arguments, for example::

    (x: Int, y: String) -> Bool

As in TypeScript, the argument names if provided are just there for documentation
and are not part of the type itself.

Scala
~~~~~

`Scala <https://docs.scala-lang.org/tour/higher-order-functions.html>`_
uses the ``=>`` arrow for function types. Other than that, their syntax is
the same as the one we are proposing, for example::

    (Int, String) => Bool

Scala, like Python, has the ability to provide function arguments by name.

> Funciton types can optionally include names, for example::

Reviewer 3: ```suggestion
Function types can optionally include names, for example::
```

    (x: Int, y: String) => Bool

Unlike in TypeScript and Kotlin, these names are part of the type if
provided - any function implementing the type must use the same names.
This is similar to the extended syntax proposal we described in our
`Rejected Alternatives`_ section.

The ML Language Family
~~~~~~~~~~~~~~~~~~~~~~

Languages in the ML family, including `F#
<https://docs.microsoft.com/en-us/dotnet/fsharp/language-reference/fsharp-types#syntax-for-types>`_,
`OCaml
<https://www2.ocaml.org/learn/tutorials/basics.html#Defining-a-function>`_,

> and ``Haskell <https://wiki.haskell.org/Type_signature>`_, all use

Reviewer 3: ```suggestion
and `Haskell <https://wiki.haskell.org/Type_signature>`_, all use
```
(This markup bug looks serious, and I'm going to commit it.)

  of the added features.
- The group that debated these proposals was split down the middle
  about whether these changes are even desirable:

  - On the one hand, they make callable types more expressive. On the
    other hand, they could easily confuse users who have not read the
    full specification of callable type syntax.
  - We believe the simpler syntax proposed in this PEP, which
    introduces no new semantics and closely mimics syntax in other
    popular languages like Kotlin, Scala, and TypesScript, is much
    less likely to confuse users.

- We intend to implement the current proposal in a way that is
  forward-compatible with the more complicated extended syntax. If the
  community decides after more experience and discussion that we want
  the additional features, they should be straightforward to propose
  in the future.
- We realized that because of overloads, it is not possible to replace
  all need for Callback Protocols even with an extended syntax. This
  makes us prefer proposing a simple solution that handles most use
  cases well.

We confirmed that the current proposal is forward-compatible with
extended syntax by

> `implementing<https://github.com/stroxler/cpython/tree/callable-type-syntax--extended>`_

Reviewer 3: ```suggestion
`implementing <https://github.com/stroxler/cpython/tree/callable-type-syntax--extended>`_
```

static analysis tooling, and code review.

One of the types `defined by PEP 484
<https://www.python.org/dev/peps/pep-0484/#callable>`_ in the is
``typing.Callable``, which describes a callable value such as a
function or a method. It takes two parameters as inputs but with the
first parameter being either a placeholder like ``...`` or a list of
types. For example:

> - ``Callable[..., int]`` indicates a funciton with arbitrary

Reviewer: ```suggestion
- ``Callable[..., int]`` indicates a function with arbitrary
```
